README.md
# appositron

Tools and utiities for the Apposition Muon Camera (AMC) project.
See https://hackaday.io/project/87925-apposition-muon-camera

## Installation

Download from https://github.com/PaulSchulz/amc

## Usage

This program is a 'clojure' application. The easiest way to run this
code with all of it's dependancies is to use the 'leiningen' tool.

Leiningen can be downloaded from here: https://leiningen.org

Run the program (and download dependancies when first run)

    $ lein repl

FIXME: explanation

    $ java -jar apposition-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

(Nothinbg to see here. Move along.)


## License

Copyright © 2018 GPL v3

Distributed under the GNU Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This change log follows the conventions of
[keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-03-16
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-03-16
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/PaulSchulz/amc/compare/0.1.1...HEAD
[0.1.1]: https://github.com/PaulSchulz/amc/compare/0.1.0...0.1.1


test/appositron/core_test.clj
(ns appositron.core-test
  (:require [clojure.test :refer :all]
            [appositron.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/apponmutron/core_test.clj
(ns apponmutron.core-test
  (:require [clojure.test :refer :all]
            [apponmutron.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/appositron/analemma.clj
(ns appositron.analemma
  (:use [analemma.charts :only [emit-svg xy-plot add-points]]
	[analemma.svg]
	[analemma.xml]
	[clojure.java.io :only [file]]))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; ANALEMMA CHART
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; http://www.wsanford.com/~wsanford/exo/sundials/analemma_calc.html
(def analemma-data
     [[-15.165	-23.07]
      [-17.016	-22.70]
      [-19.171	-22.08]
      [-21.099	-21.27]
      [-22.755	-20.30]
      [-24.107	-19.16]
      [-25.446	-17.33]
      [-25.914	-16.17]
      [-26.198	-14.62]
      [-26.158	-12.96]
      [-25.814	-11.21]
      [-25.194	-9.39]
      [-24.520	-7.89]
      [-23.708	-6.37]
      [-22.529	-4.42]
      [-21.205	-2.45]
      [-19.777	-0.48]
      [-18.289	1.50]
      [-16.185	4.24]
      [-15.009	5.78]
      [-13.605	7.66]
      [-12.309	9.49]
      [-11.153	11.26]
      [-10.169	12.94]
      [-9.250	14.85]
      [-8.811	16.04]
      [-8.469	17.43]
      [-8.364	18.69]
      [-8.493	19.83]
      [-8.847	20.82]
      [-9.685	21.96]
      [-10.317	22.47]
      [-11.231	22.96]
      [-12.243	23.28]
      [-13.308	23.43]
      [-14.378	23.41]
      [-15.599	23.16]
      [-16.339	22.86]
      [-17.139	22.33]
      [-17.767	21.64]
      [-18.191	20.80]
      [-18.387	19.81]
      [-18.253	18.20]
      [-17.956	17.17]
      [-17.361	15.78]
      [-16.529	14.28]
      [-15.474	12.68]
      [-14.221	11.01]
      [-12.183	8.54]
      [-10.901	7.07]
      [-9.212	5.20]
      [-7.462	3.29]
      [-5.693	1.36]
      [-3.946	-0.59]
      [-1.938	-2.93]
      [-0.686	-4.48]
      [0.742	-6.39]
      [1.982	-8.28]
      [2.993	-10.11]
      [3.742	-11.88]
      [4.290	-14.23]
      [4.318	-15.49]
      [4.044	-16.97]
      [3.420	-18.33]
      [2.446	-19.55]
      [1.135	-20.63]
      [-0.852	-21.71]
      [-2.398	-22.29]
      [-4.538	-22.86]
      [-6.855	-23.24]
      [-9.286	-23.42]
      [-11.761	-23.41]
      [-14.691	-23.14]])

(defn analemma [filename]
  (spit filename
	(emit-svg
	 (-> (xy-plot :xmin -30 :xmax 10,
		      :ymin -30 :ymax 30
		      :height 500 :width 500)
	     (add-points analemma-data)))))


(defn analemma-logo [filename]
  (spit filename
	(emit
	  (svg
	   (apply group
		  (-> (text "Analemma")
		      (add-attrs :x 120 :y 60)
		      (style :fill #"000066"
			     :font-family "Garamond"
			     :font-size "75px"
			     :alignment-baseline :middle))
		  (for [[x y] analemma-data]
		    (circle (translate-value x -30 5 0 125)
			    (translate-value y -25 30 125 0)
			    2 :fill "#000066")))))))

(defn rotating-analemma-logo [filename]
  (spit filename
	(emit
	 (svg
	  (-> (image "file:images/analemma-logo.svg"
		     :width 500 :height 700)
	      (animate-transform :begin 0
					:dur 20
					:type :rotate
					:from "0 200 150"
					:to "360 200 150"
					:repeatCount :indefinite))))))


(defn query-and-transform-analemma [filename]
  (let [logo (parse-xml (slurp (file "images/analemma-logo.svg")))
	red-analemma (-> (concat [:g] (filter-xml logo [:g :circle]))
			 (transform-xml [:circle] #(add-attrs % :fill "#FF0000")))]
    (spit filename
	 (emit
	  (svg
	   (-> red-analemma
	       (animate-transform :begin 0
				  :dur 20
				  :type :rotate
				  :from "0 100 100"
				  :to "360 100 100"
				  :repeatCount :indefinite)))))))


src/appositron/core.clj
(ns appositron.core
  (:gen-class)
  (:use [analemma svg charts xml]
        [tikkba swing dom]
        tikkba.utils.xml)
  (:require [tikkba.utils.dom :as dom]))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))

(def page-width  600)
(def page-height 600)

;; Array of azimuth, declination
;;  Multiple domains
(def aperture
  [[  0.0 10.0]
    [ 10.0 10.0]
    [ 20.0 10.0]
    [ 30.0 10.0]
    [ 40.0 10.0]
    [ 50.0 10.0]
    [ 60.0 10.0]
    [ 70.0 10.0]
    [ 80.0 10.0]
    [ 90.0 10.0]
    [100.0 10.0]
    [110.0 10.0]
    [120.0 10.0]
    [130.0 10.0]
    [140.0 10.0]
    [150.0 10.0]
    [160.0 10.0]
    [170.0 10.0]
    [180.0 10.0]
    [190.0 10.0]
    [200.0 10.0]
    [210.0 10.0]
    [220.0 10.0]
    [230.0 10.0]
    [240.0 10.0]
    [250.0 10.0]
    [260.0 10.0]
    [270.0 10.0]
    [280.0 10.0]
    [290.0 10.0]
    [300.0 10.0]
    [310.0 10.0]
    [320.0 10.0]
    [330.0 10.0]
    [340.0 10.0]
    [350.0 10.0]
    ])

;; Aperature channels
(def channel-colours
  ["#ff0000"  ; Red
   "#00ff00"  ; Green
   "#0000ff"  ; Blue
   ])

;;  Chanels Array
;;  -> Aperture Segment Array
;;     -> Segment
;;        -> Boundry Points
(def aperture-channels
  [[   0.0  20.0    5.0  21.0]
   [   5.0  21.0   10.0  22.0]
   [  10.0  22.0   30.0  90.0]
   [  30.0  90.0   25.0  90.0]
   [  25.0  90.0   20.0  90.0]
   [  20.0  90.0   15.0  90.0]
   [  15.0  90.0   10.0  90.0]
   [  10.0  90.0    5.0  90.0]
   [   5.0  90.0    0.0  90.0]
   [   0.0  90.0   -5.0  90.0]
   [  -5.0  90.0  -10.0  90.0]
   [ -10.0  90.0  -15.0  90.0]
   [ -15.0  90.0  -20.0  90.0]
   [ -20.0  90.0  -25.0  90.0]
   [ -25.0  90.0  -30.0  90.0]
   [ -30.0  90.0  -10.0  22.0]
   [ -10.0  22.0   -5.0  21.0]
   [  -5.0  21.0    0.0  20.0]
   ]
)

(defn transform-aperture [aperture]
  (map (fn [[a b]] [(* b (Math/cos a))
                    (* b (Math/sin a))]) aperture)
  )

(defn deg2rad [degrees]
  (* (/ degrees 360) Math/PI 2)
)

(defn calc-x [r t]
  (+ (* (+ (* (* r (Math/cos (deg2rad t))))) (/ 300 90)) 300)
  )

(defn calc-y [r t]
  (+ (* (+ (* (* r (Math/sin (deg2rad t))))) (/ 300 90)) 300)
  )

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn create-aperture-full-svg
  ""
  []
  (let [center-x (/ page-width 2)
        center-y (/ page-height 2) 
        TwoPi (* Math/PI 2)
        ]
  (svg
   (apply group
          (-> (rect 0 0 page-width page-height :id "rect0")
              (style :fill "white" :stroke "blue" :stroke-width 0))

          (-> (text {:x 10 :y 20} "Aperture")
              (style :font-family "Courier" :font-size 16 :fill "blue"))              

          (-> (circle (/ page-width 2) (/ page-height 2) 300 :id "circle")
              (style  :fill :none :stroke "blue" :stroke-width 1))
          (-> (circle (/ page-width 2) (/ page-height 2) 200 :id "circle")
              (style  :fill :none :stroke "blue" :stroke-width 1))
          (-> (circle (/ page-width 2) (/ page-height 2) 100 :id "circle")
              (style  :fill :none :stroke "blue" :stroke-width 1))
          
          (-> (path [:M [center-x center-y]
                     :l [300 0]
                     :M [center-x center-y]
                     :l [-300 0]
                     :M [center-x center-y]
                     :l [0 -300]
                     :M [center-x center-y]
                     :l [0 300]
                     
                     ])
              (style  :fill :none :stroke "blue" :stroke-width 1))
          (-> (text {:x (+ center-x 78)
                     :y (+ center-y 16)}
                    "30")
              (style :font-family "Arial" :font-size 16 :fill "blue"))
          (-> (text {:x (+ center-x 178)
                     :y (+ center-y 16)}
                    "60")
              (style :font-family "Arial" :font-size 16 :fill "blue"))
          (-> (text {:x (+ center-x 278)
                     :y (+ center-y 16)}
                    "90")
              (style :font-family "Arial" :font-size 16 :fill "blue"))              

          (for [[x y m n] aperture-channels]
            (let [a1 (calc-x y x)
                  a2 (calc-y y x)
                  b1 (calc-x n m)
                  b2 (calc-y n m)
                  ]
;            (circle a1 a2 2 :fill "#000000")))
              (-> (path [:M [a1 a2]
                         :L [b1 b2]])
              (style :stroke "blue" :stroke-width 1))

             )
            )

          )
   )))

(defn create-crosssection-svg
  ""
  []
  (let [center-x (/ page-width 2)
        center-y (/ page-height 2) 
        TwoPi (* Math/PI 2)
        ]
  (svg
   (-> (rect 0 0 page-width page-height :id "rect0")
       (style :fill "white" :stroke "blue" :stroke-width 0))

   ;; (-> (circle (/ page-width 2) (/ page-height 2) 300 :id "circle")
   ;;     (style  :fill :none :stroke "blue" :stroke-width 1))
   ;; (-> (circle (/ page-width 2) (/ page-height 2) 200 :id "circle")
   ;;     (style  :fill :none :stroke "blue" :stroke-width 1))
   ;; (-> (circle (/ page-width 2) (/ page-height 2) 100 :id "circle")
   ;;     (style  :fill :none :stroke "blue" :stroke-width 1))

   (-> (text {:x 10 :y 20} "Crosssection - Declination:   0.0 Aximuth:   0.0")
       (style :font-family "Courier" :font-size 16 :fill "blue"))              

   (-> (path [:M [center-x center-y]
              :l [300 0]
              :M [center-x center-y]
              :l [-300 0]
              :M [center-x center-y]
              :l [0 -300]
              :M [center-x center-y]
              :l [0 300]

              ])
       (style  :fill :none :stroke "blue" :stroke-width 1))
   (-> (text {:x (+ center-x 78)
              :y (+ center-y 16)}
             "10")
       (style :font-family "Arial" :font-size 16 :fill "blue"))
   (-> (text {:x (+ center-x 178)
              :y (+ center-y 16)}
             "20")
       (style :font-family "Arial" :font-size 16 :fill "blue"))
   (-> (text {:x (+ center-x 278)
              :y (+ center-y 16)}
             "30")
       (style :font-family "Arial" :font-size 16 :fill "blue"))              

   )))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn write-svg
  [create-svg file-svg]
  ;; Converts the SVG representation to a XML Document
  ;; and writes it to a file
  (let [doc (svg-doc (create-svg))]
    (dom/spit-xml file-svg doc
                  :indent "yes"
                  :encoding "UTF8")))

(write-svg create-aperture-full-svg "aperture-full.svg")
(write-svg create-crosssection-svg  "crosssection.svg")
